### ISSUE TITLE
ergonomics & encodings

### ISSUE THREAD

**Issue by cmyr** [tafia/quick-xml#158]
Hi there,

I've just been playing around with quick-xml to handle reading and writing of [Glyph Interchange Format](http://unifiedfontobject.org/versions/ufo3/glyphs/glif/#contour) files, and I had some observations I wanted to share.

## `[u8]`/`Cow<[u8]>` vs `Cow<str>`

My understanding is that `quick-xml` attempts to do the least work possible, and to this end exposes byte slices by default. I can understand the reasoning (don't allocate unless the user has asked for it explicitly) but there are two things I keep thinking about, here:

- _utf-8 is common_: the majority of XML I've been dealing with is encoded in utf-8. In this case, we should be able to provide direct access to string slices, allocating only if we need to escape something.

- _In a non-utf8 encoding, it is hard to use bytes correctly_: exposing bytes directly encourages users to do things like `if attr.key == b"mykey"` (which is used in the readme). This is fine with an ascii-compatible encoding like utf-8 or Shift JIS, but will fail unexpectly on, say, utf-16; and utf-16 is an explicitly supported encoding according to the [xml spec](https://www.w3.org/TR/REC-xml/#charencoding).

## API thoughts

I haven't thought about this too too much, but I do worry that the current behaviour has worse ergonomics than necessary in the general (utf-8) case, while also making it easy to write bugs in the exceptional (utf-16) case.

One possible alternative I could imagine would be having separate `Event` and `EventRaw` types; `Event` would work with `Cow<str>`, which would be slices of the underlying buffer where possible (when the encoding is utf-8) and which would be decoded and allocated during parsing, otherwise. `EventRaw` would work like the current `event`.

Another option would be to have add some new type, `XmlSlice`, which would include a lazily-allocating `to_str` method, while also exposing access to the underlying bytes and even the encoding; this type would be used everywhere that `[u8]` or `Cow<[u8]>` is, currently. This could also include information like the location of this text in the original document.

## thanks

I want to clarify that this is not a feature request or a concrete proposal for a new API; I just wanted to voice these thoughts and hear if anyone had anything to add. It may be that what I'm describing should just be an alternative crate, or that there are other design constraints I'm not aware of.

Thank you for the time you've put into this code so far!

**Comment by tafia** [comment]
Thanks for reaching out, these kind of feedbacks are very useful!

I agree this is not super ergonomic but in practice I don't feel this is really an issue. This is very subjective of course and i'd be more than happy to improve the current situation. Just to be sure we're on the same page we want to avoid both converting (checking) to str AND decoding when possible (we'd rather encode the matching keyword once, most of the time at compile time than decode all events).  

Having 2 types of events is definitely a good idea and could help. I prefer the `XmlSlice` idea because it is lazier and it doesn't hide the fact that the data is not guaranteed to be utf8. XmlSlice could deref to `[u8]` and could either implement some `TryFrom` or have dedicated `to_str` methods.

**Comment by cmyr** [comment]
Okay, glad to hear that you're receptive to this. I'm not sure if it's something I'm going to dig too deeply into, but if I do I'll propose a more thorough design first.

Can you clarify precisely what you mean by checking / decoding? Are you talking only about the source xml data, or are you also thinking about the constants that keys are matched against?

**Comment by argv-minus-one** [comment]
>we want to avoid both converting (checking) to str AND decoding when possible

Rust performs those checks for a reason: failing to do so [is a security vulnerability.](https://unicode.org/reports/tr36/#UTF-8_Exploit) Skipping decoding also makes it impossible to support ASCII-incompatible encodings like UTF-16 (which is [required by the XML spec](https://www.w3.org/TR/xml/#charsets), by the way).

A correct and secure XML parser must always decode.

**Comment by tafia** [comment]
Right. I don't see why these checks wouldn't be performed here as well, thus I don't see how it'd impact security.

Granted it is a manual process but an absolutely **mandatory** one:

```rust
Event::Start(utf16_event) => {
    let utf8_name =  reader.decode(utf16_event.local_name())?;
    match utf8_name {
         "utf8 name1" => { ... }
         "utf8 name2" => { ... }
    }        
```

Please correct me if I'm wrong.

As far as I know, quick-xml is the only one which actually lets you deal with non-utf8 data at all!

The way the api is done is to provide the developer who knows the xml he is dealing with to avoid escaping/decoding it when possible, which is very common.

**Comment by scottlamb** [comment]
> The way the api is done is to provide extra perf to developers who know the xml they are dealing with, which is very common in my experience.

I'm interested to know what your design goal is (both for this issue and general). Do you always want `quick-xml` to default to (or be configurable for) maximum speed at all costs? or is it acceptable to make decisions that degrade performance a little (say, up to 5%, for the sake of having a number) for other goals: ergonomics, specification compliance, etc?

`quick-xml` is an order of magnitude faster than `xml-rs`; personally, if it were 5% slower than it is now but improved those other qualities, I'd still call it quick, and I'd be really happy.

Besides ergonomics, I think the current encoding design doesn't work. In particular, because of the assumption here that decoding is just something the caller does on text nodes as needed, `quick-xml` only can handle UTF-8 and single-byte encodings that are compatible with ASCII (such as ISO-8859-1):
*   It doesn't use external encoding information and doesn't detect the encoding based on the opening bytes (as in [Appendix F](https://www.w3.org/TR/xml/#sec-guessing) of the spec), so `read_question_mark` won't take its `&buf[1..4] == b"xml"` path and will never find the right encoding.
*    Even if it did pick the right encoding, subsequent stuff likewise seems to get compared and sliced assuming its encoding matches ASCII, so UTF-16, Shift-JIS, EBCDIC, etc. won't work.

I see there's issue #322 open about UTF-16 not working, and a test of UTF-16 is `#[ignore]`d.

I think the most straightforward approach would be (as well as fixing the detection) to decode non-UTF-8 encodings into a buffer and feed that into the XML parsing. Then everything afterward is UTF-8. And then it can give `str` to callers at no additional cost on those encodings (assuming `encoding-rs`'s output is valid), and just for the cost of `std::str::from_utf8` on UTF-8. I believe a fair amount of effort has gone into optimizing `std::str::from_utf8`, so I'm hopeful the performance degradation would be minimal.

btw:

> Coming back to the security examples, quick-xml will not try to decode anything per default and will not try to delete or replace any character. True, it will NOT break per default if the data is invalid but I believe it will not let you insert anything malicious.

I'm not sure I agree with this. In particular, I suspect it's possible to do "interesting" things via non-canonical UTF-8 and the like. For example, I think the duplicate attribute detection works on the raw bytes now, so by not validating the UTF-8 it's possible to defeat it. Then maybe two implementations differ on which one they use, bypassing some validation logic or something. There are people who are really good at coming up with crazy scenarios vaguely like this that result in actual security holes in some system or another. I'm not really one of them but I'm doing my best to think like them.

**Comment by dralley** [comment]
I've been thinking about this a bit lately (and writing my thoughts on this [PR](https://github.com/tafia/quick-xml/pull/415#issuecomment-1180691911=)) so I will copy them here where they are more relevant.

I've come to agree with the OP and also, in particular, this comment 

>I think the most straightforward approach would be (as well as fixing the detection) to decode non-UTF-8 encodings into a buffer and feed that into the XML parsing. Then everything afterward is UTF-8. And then it can give str to callers at no additional cost on those encodings (assuming encoding-rs's output is valid), and just for the cost of std::str::from_utf8 on UTF-8. I believe a fair amount of effort has gone into optimizing std::str::from_utf8, so I'm hopeful the performance degradation would be minimal.

To start with, I'd like to provide some data on just how fast UTF-8 validation can be.  If you look at the benchmarks in https://github.com/rusticstuff/simdutf8, even on years old hardware (Zen 2) it is possible to validate utf-8 compatibility on large blocks of text at a speed of >100gb per second.  The standard library doesn't support this yet (there is an open issue to port the work to stdlib) but in any case, even the standard library `std::str::from_utf8()` can manage ~30gb per second.

One thing you should notice though is that this only applies to large blocks of data.  Validating small strings is much **much** slower, partly because of cache effects, partly because SIMD won't benefit as much as it will with larger inputs.  So if you take the assumption that users often need to work with a significant portion of the XML as a UTF-8 encoded type (`String` or `&str`), pre-validating the buffer UTF-8 could easily be an overall performance win for many users, rather than being "more work".

Another issue is that if the user does need to support multiple encodings, with the current API (assuming utf-16 support were eventually implemented), they would need to decode things everywhere, which means a lot of small allocations.  The performance cost of those allocations is likely to be as bad or worse than simply decoding everything up-front and working with UTF-8.

So the performance case against this is actually not very clear cut.

Combine this with some of the other points brought up: 

* the testing burden of testing parsing code against every encoding
* the maintainence burden of maintaining parsing code that works for every encoding
* the maintainence burden of every struct (`Bytes*`, `Attribute` and so forth) needing to know about decoding
* the user frustration of needing to change all of your code if you used the simple utf-8 methods and then suddenly need to parse a utf-16 document and nothing works anymore without changing the code
* the user burden of needing to convert to `&str` in many places
* the API bloat where many functions need an `_and_decode` equivalent

And I think, it actually does make sense to follow the same path as other XML parsers including [libxml2](https://gitlab.gnome.org/GNOME/libxml2/-/wikis/Encodings-support#the-internal-encoding-how-and-why) by decoding everything up-front into a buffer which can always be assumed to be valid UTF-8.

**Comment by dralley** [comment]
I also ran the `encoding_rs` benchmark suite to get an idea of how fast encoding / decoding UTF-16 is.  It's good enough that I wouldn't expect more than about 30% overhead compared to UTF-8 - and again, if the user does a lot of decoding themselves it would flip the other way.

**Comment by lukaslihotzki** [comment]
About ergonomics: Is it possible to construct a `BytesEnd` corresponding to an existing `BytesStart`? This would be easy with `BytesEnd::wrap`, but this isn't public. `BytesEnd::new` takes a `str`, that I would have to create manually from the `[u8]` of `BytesStart`, and this would fail on non-utf8 encodings. I can enclose the name in `<…></…>` and create a new `Reader`, but this is extremely unergonomic.

Based on the previous comment of @dralley and for implementation of UTF-16 with reasonable effort, I would prefer to convert to or validate as UTF-8 before parsing. Maybe quick-xml could offer both strategies: For example, a const bool `UTF8` could be added to the XML Reader and Writer, and the XmlSlice type discussed here. If `UTF8 = false`, readers and writers consume/produce `[u8]`, and the XmlSlice needs to validate on conversion to `str`. If `UTF8 = true`, readers only accept `str` as inputs, writers offer `str` output, and XmlSlice can convert to `str` without validation.

**Comment by Mingun** [comment]
> About ergonomics: Is it possible to construct a `BytesEnd` corresponding to an existing `BytesStart`?

Yes. [BytesStart::to_end()](https://docs.rs/quick-xml/latest/quick_xml/events/struct.BytesStart.html#method.to_end).

**Comment by eirnym** [comment]
Following workaround for UTF-16 is bulky and maybe somewhat slow, but it works at least. I'd love to have this implemented somehow inside the library (feature-gated) to make it more optimal and not copy-paste it everywhere.

Some notes: 

* optimizations are welcome
* This doesn't fix `quick-xml` byte parsing, just a workaround for `UTF-16` encoding parsing.
* `from_utf16be` and `from_utf16le` can be replaced by `String::from_utf16be` and `String::from_utf16le` respectively when they are out.
* `anyhow` is just for this POC
* `byteorder` is not so strictly required, and may be replaced with standard library
* `std::io::Cursor` may be replaced by `nostd` variant.
* `LocalEncoding` is an enum with 3 elements, and `detect_utf16` is a copy of `detect_encoding` function from this library returning an enum. Copy is to skip `encoding` feature`. 
* Curiously enough, it works even `encoding` feature is enabled.

```rust
use anyhow::Result;
use byteorder::{BigEndian, LittleEndian, ReadBytesExt};
use serde::de::DeserializeOwned;

use crate::encoding_detect::LocalEncoding;
use crate::encoding_detect::detect_utf16;

pub fn parse_xml<'de, Type>(data: &[u8]) -> Result<Type>
where
    Type: DeserializeOwned,
{
    match detect_utf16(data) {
        Some(LocalEncoding::Utf16be) => {
            let value = from_utf16be(data)?;
            quick_xml::de::from_str::<Type>(&value).map_err(|err| err.into())
        }
        Some(LocalEncoding::Utf16le) => {
            let value = from_utf16le(data)?;
            quick_xml::de::from_str::<Type>(&value).map_err(|err| err.into())
        }
        _ => {
            let io: std::io::Cursor<&[u8]> = std::io::Cursor::new(data);
            quick_xml::de::from_reader::<std::io::Cursor<&[u8]>, Type>(io).map_err(|err| err.into())
        }
    }
}

fn from_utf16le(data: &[u8]) -> Result<String> {
    let mut source = std::io::Cursor::new(data);
    let mut dst: Vec<u16> = std::vec::from_elem(0, data.len() / 2);
    source.read_u16_into::<LittleEndian>(&mut dst)?;
    String::from_utf16(&dst).map_err(|err| err.into())
}

fn from_utf16be(v: &[u8]) -> Result<String> {
    let mut rdr = std::io::Cursor::new(v);
    let mut output: Vec<u16> = Vec::with_capacity(v.len() / 2);
    rdr.read_u16_into::<BigEndian>(&mut output)?;
    String::from_utf16(&output).map_err(|err| err.into())
}

```

**Comment by dralley** [comment]
I am working on a resolution to this issue: https://github.com/tafia/quick-xml/pull/963

General principles:

* `Reader` validates and returns `utf-8` (apart from `.stream()`, which acts as an escape hatch) for all XML component types
* `DecodingReader` acts as a wrapper which users can pass to transcoding documents to UTF-8 before passing them to `Reader`
* All need for "decoding" disappears from the API, apart from pre-processing

**Comment by dralley** [comment]
(Apologies, I misclicked, didn't mean to close)

**Comment by dralley** [comment]
What I want this to look like, ultimately:

If the "decoding" feature is enabled:

* `DecodingReader` is built in-to `Reader` and validates or transcodes input automatically
* Therefore, the parser can use `std::str::from_utf8_unchecked()` at XML element boundaries, eliminating all decoding overhead from small strings - the decoding loop therefore remains a tight loop with good data/cache locality and maximum SIMD benefits from operating over a large buffer
* The user need not do anything manually to handle alternate document encodings, they are decoded transparently

If the "decoding" feature is not enabled:

* `Reader` uses `std::str::from_utf8()` (or similar) at XML element boundaries, doing the validation lazily and allowing the consumption of non-UTF-8 data embedded within the XML using `BinaryStream` / `stream()`, which would bypass the validation.
* The user still does not need to do anything manually to decode the document, but the portions of the input processed as XML must be UTF-8

Using a similar approach, we can do XML end of line normalization and line / column tracking within a preprocessing layer and eliminate that burden from the user as well.

**Comment by Mingun** [comment]
> If the "decoding" feature is not enabled:
> * `Reader` uses `std::str::from_utf8()` (or similar) at XML element boundaries, doing the validation lazily and allowing the consumption of non-UTF-8 data embedded within the XML using `BinaryStream` / `stream()`, which would bypass the validation.

Do you mean, that if "decoding" will be enabled, we will not have `BinaryStream` or it will always return UTF-8 data and it become useless? If so, that will violate the rules about "additivity" of features. This is undesirable, although possible if there are no better alternatives.

**Comment by dralley** [comment]
>or it will always return UTF-8 data and it become useless?

This.  There's no particular reason to disable the API entirely, it would just return post-validated bytes.  I don't think it's necessarily "useless" but I agree that the original motivation (embedded binary blobs within an XML document) would not apply.

**Comment by dralley** [comment]
Alternatively we could flip the options around, make `decoding` the default behavior, and opt-in to lazy UTF-8 validation, which would then enable the `stream()` API and make the features additive again.

edit: actually that just breaks things in the opposite direction, because you're removing the ability to decode by enabling a feature.

TBH I'm not sure I consider it a violation of additivity in the first place though.  If the feature is "transcode my inputs" then that necessarily means the inputs must be transcodable.  The limitation and the feature are tied together fundamentally, and the same in reverse.  

No other general purpose XML parser that I know of has a `stream()` equivalent, so if it makes the features a bit awkward, that might just be the cost of admission.

Or maybe we could have a runtime switch that disables the transparent decoding and enables the lazy encoding.  I guess that's an option.

**Comment by dralley** [comment]
Or, we could make "eager decoding" (e.g. automatic `DecodingReader`) be configurable at runtime (at construction - not arbitrarily during execution).

If a user needs to use `stream()` in conjunction with un-decodable content, they could configure the reader to use lazy decoding (wherein the parser would decode on a per-event basis - but still transparently).

**Comment by dralley** [comment]
I'm going to close this as complete (the API is now `str` / `String` based) and file a new issue to improve the integration of DecodingReader with `Reader`.

**Comment by dralley** [comment]
An example of some of the ergonomics improvements now that the `Reader` uses utf-8 types natively https://github.com/dralley/rpmrepo_metadata/pull/22/changes
